homeassistant/components/tahoma/coordinator.py

```diff
@@ -0,0 +1,150 @@
+"""Helpers to help coordinate updates."""
+from datetime import timedelta
+import logging
+from typing import Dict, List, Optional, Union
+
+from aiohttp import ServerDisconnectedError
+from pyhoma.client import TahomaClient
+from pyhoma.enums import EventName, ExecutionState
+from pyhoma.exceptions import (
+    BadCredentialsException,
+    MaintenanceException,
+    NotAuthenticatedException,
+    TooManyRequestsException,
+)
+from pyhoma.models import DataType, Device, State
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+TYPES = {
+    DataType.NONE: None,
+    DataType.INTEGER: int,
+    DataType.DATE: int,
+    DataType.STRING: str,
+    DataType.FLOAT: float,
+    DataType.BOOLEAN: bool,
+}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TahomaDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching TaHoma data."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        *,
+        name: str,
+        client: TahomaClient,
+        devices: List[Device],
+        update_interval: Optional[timedelta] = None,
+    ):
+        """Initialize global data updater."""
+        super().__init__(
+            hass,
+            logger,
+            name=name,
+            update_interval=update_interval,
+        )
+
+        self.data = {}
+        self.original_update_interval = update_interval
+        self.client = client
+        self.devices: Dict[str, Device] = {d.deviceurl: d for d in devices}
+        self.executions: Dict[str, Dict[str, str]] = {}
+
+        _LOGGER.debug(
+            "Initialized DataUpdateCoordinator with %s interval.", str(update_interval)
+        )
+
+    async def _async_update_data(self) -> Dict[str, Device]:
+        """Fetch TaHoma data via event listener."""
+        try:
+            events = await self.client.fetch_events()
+        except BadCredentialsException as exception:
+            raise UpdateFailed("invalid_auth") from exception
```

Comment:
Please write a statement in the exception message, not a single word.

Reply:
Fixed.